src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/RenameEndpoint.cs

```diff
@@ -88,6 +100,46 @@ internal class RenameEndpoint : IRenameEndpoint
                 return null;
             }
 
+            var razorEdits = await TryGetRazorComponentRenameEditsAsync(request, requestDocumentSnapshot, codeDocument, cancellationToken).ConfigureAwait(false);
+
+            if (razorEdits is not null)
+            {
+                return razorEdits;
+            }
+
+            // If we're not doing single server rename, then we're done. C# and Html will be handled by the RenameHandler in the HtmlCSharp server.
+            if (!_languageServerFeatureOptions.SingleServerRenameSupport)
+            {
+                return null;
+            }
+
+            var sourceText = await documentContext.GetSourceTextAsync(cancellationToken).ConfigureAwait(false);
+            if (!request.Position.TryGetAbsoluteIndex(sourceText, _logger, out var absoluteIndex))
+            {
+                return null;
+            }
+
+            var projection = await documentContext.GetProjectionAsync(absoluteIndex, _documentMappingService, cancellationToken).ConfigureAwait(false);
+
+            // If the language is Razor then the downstream servers won't know how to handle it anyway
+            if (projection.LanguageKind == Protocol.RazorLanguageKind.Razor)
+            {
+                return null;
+            }
+
+            var delegatedParams = new DelegatedRenameParams(
+                documentContext.Identifier,
+                projection.Position,
+                projection.LanguageKind,
+                request.NewName);
+            var delegatedRequest = await _languageServer.SendRequestAsync(LanguageServerConstants.RazorRenameEndpointName, delegatedParams).ConfigureAwait(false);
+            var delegatedResponse = await delegatedRequest.Returning<WorkspaceEdit?>(cancellationToken).ConfigureAwait(false);
+
+            return delegatedResponse;
```

Comment:
I feel like this depends on the impl detail of the VS LSP server to eagerly remap locations. I'm not sure that's also true of O#'s impl. We may want to dig through the response and see if there's any locations that we know how to remap.

Reply:
Yeah, I will add remapping here. I actually started doing it, but removed it when i noticed it wasn't being used. It's a bit annoying to do at this layer, interestingly.

Reply:
> It's a bit annoying to do at this layer, interestingly.

Oh wow why's that?